include/circt/Dialect/MSFT/MSFTConstructs.td

```diff
@@ -67,3 +67,48 @@ def ChannelOp: MSFTOp<"constructs.channel",
     $input $clk $sym_name `(` $defaultStages `)` attr-dict `:` type($input)
   }];
 }
+
+// Linear, pipelineable datapath.
+def LinearOp : MSFTOp<"hlc.linear", [
+    SingleBlockImplicitTerminator<"OutputOp">,
+    IsolatedFromAbove
+  ]> {
+  let summary = "Model of a linear datapath which can be arbitrarily pipelined";
+  let description = [{
+    Defines a feed-forward datapath which can be scheduled into a pipeline.
+    Due to the feed-forwardness, the inner region is NOT a graph region.
+    Internally, only combinational operations (`comb`, `msft`, `hw`) are allowed.
+    The operation expects exactly one i1-typed port to be tagged with a
+    `msft.clock` attribute.
+
+    Example:
+    ```mlir
+    msft.module @foo(%in0 : i32, %in1 : i32, %in2 : i32, %clk : i1) -> (out: i32) -> {
+      %0 = msft.hlc.linear(%a = %in0, %b = %in1, %c = %in2) clock %clk (i32, i32, i32) -> (i32) {
```

Comment:
Is there no way to re-use the operand names so as to avoid the `(%a = %in0, %b = %in1, %c = %in2)`? Ideally, we'd be able to use values in parent blocks directly (which I thought was allowed) but my experience has been that they're not.

Reply:
This whole schema is due to marking the operation as `IsolatedFromAbove`. I've debated with myself whether we should allow capturing (by reference) values from the enclosing scope. That would obviously remove this, since `%in0, %in1,...` could just be referenced inside the datapath. My motivation for `IsolatedFromAbove` (and thus the explicit passing-in of arguments) was to be intentional about what - and where - values are passed into the pipeline.

If we decide to create a PyCDE API for this similar to the `SystolicArray` construct, then `IsolatedFromAbove` makes sense:
```python
@generator
def generate(ports):
   def datapath(in0, in1, in2):
      return (in0 * in1) + in2
   
   ports.out = Linear(datapath, ports.in0, ports.in1, ports.in2)
```

However, if we expect to do some `with` syntax, it should probably be removed, e.g.:
```python
@generator
def generate(ports):
  mac = Linear(types.i32) # Predeclare with return type
  with mac:
    mac.out0 = (ports.in0 + ports.in1) + ports.in2
  
  ports.out = mac.out0
```

Reply:
Sounds like we should pivot to the `IsolatedFromAbove` approach, then!. Note that this will require an analysis to infer externally captured values when lowering to `pipeline.rtp` (fairly trivial).

Reply:
"pivot to"? You mean "pivot away from"? Or did I understand backwards?

Reply:
Pivot away from, sorry about that.